modules/backendscheduler/provider/compaction.go

```diff
@@ -0,0 +1,395 @@
+	BufferSize       int                     `yaml:"buffer_size"`
+	MeasureInterval  time.Duration           `yaml:"measure_interval"`
+	Compactor        tempodb.CompactorConfig `yaml:"compaction"`
+	MaxJobsPerTenant int                     `yaml:"max_jobs_per_tenant"`
+	Backoff          backoff.Config          `yaml:"backoff"`
+}
+
+func (cfg *CompactionConfig) RegisterFlagsAndApplyDefaults(prefix string, f *flag.FlagSet) {
+	f.DurationVar(&cfg.MeasureInterval, prefix+"backend-scheduler.compaction-provider.measure-interval", time.Minute, "Interval at which to metric tenant blocklist")
+	f.IntVar(&cfg.MaxJobsPerTenant, prefix+"backend-scheduler.max-jobs-per-tenant", 1000, "Maximum number of jobs to run per tenant before moving on to the next tenant")
+	f.IntVar(&cfg.BufferSize, prefix+"backend-scheduler.compaction-provider.buffer-size", 10, "Buffer size for compaction jobs")
+
+	// Backoff
+	f.DurationVar(&cfg.Backoff.MinBackoff, prefix+".backoff-min-period", 100*time.Millisecond, "Minimum delay when backing off.")
+	f.DurationVar(&cfg.Backoff.MaxBackoff, prefix+".backoff-max-period", 10*time.Second, "Maximum delay when backing off.")
+	f.IntVar(&cfg.Backoff.MaxRetries, prefix+".backoff-retries", 0, "Number of times to backoff and retry before failing.")
+
+	cfg.Compactor = tempodb.CompactorConfig{}
+	cfg.Compactor.RegisterFlagsAndApplyDefaults(util.PrefixConfig(prefix, "compaction"), f)
+}
+
+type CompactionProvider struct {
+	cfg    CompactionConfig
+	logger log.Logger
+
+	// Dependencies needed for compaction job selection
+	store     storage.Store
+	overrides overrides.Interface
+
+	// Scheduler calls required for this provider
+	sched Scheduler
+
+	// Dependencies needed for tenant selection
+	curPriority       *tenantselector.PriorityQueue
+	curTenant         *tenantselector.Item
+	curSelector       blockselector.CompactionBlockSelector
+	curTenantJobCount int
+}
+
+func NewCompactionProvider(
+	cfg CompactionConfig,
+	logger log.Logger,
+	store storage.Store,
+	overrides overrides.Interface,
+	scheduler Scheduler,
+) *CompactionProvider {
+	return &CompactionProvider{
+		cfg:         cfg,
+		logger:      logger,
+		store:       store,
+		overrides:   overrides,
+		curPriority: tenantselector.NewPriorityQueue(),
+		sched:       scheduler,
+	}
+}
+
+func (p *CompactionProvider) Start(ctx context.Context) <-chan *work.Job {
+	jobs := make(chan *work.Job, p.cfg.BufferSize)
+
+	go func() {
+		defer close(jobs)
+
+		measureTicker := time.NewTicker(p.cfg.MeasureInterval)
+		defer measureTicker.Stop()
+
+		level.Info(p.logger).Log("msg", "compaction provider started")
+
+		var (
+			job *work.Job
+			b   = backoff.New(ctx, p.cfg.Backoff)
+		)
+
+		reset := func() {
+			job = nil
+			b.Reset()
+		}
+
+		for {
+			select {
+			case <-ctx.Done():
+				level.Info(p.logger).Log("msg", "compaction provider stopping")
+				return
+			case <-measureTicker.C:
+				// Measure the tenants to get their current compaction status
+				p.measureTenants()
```

Comment:
i'm seeing this done here and below (line 173) should we remove the below goroutine and only do it here?